### DISCUSSION TITLE
Freeplane file chooser/file picker dialog - how to change?

### DISCUSSION THREAD

**Discussion by quickfold** [freeplane/freeplane#1823] [Help]
Hi, the Freeplane native Java file chooser is a major friction point for me. I've asked before about how to change it and @dpolivaev said that there were good reasons that Freeplane uses the Java file picker rather an the OS native file dialog: https://sourceforge.net/p/freeplane/discussion/758437/thread/02f600d6f6/

I understand that there are benefits to the native chooser, but I'm still interested in having the option of a native file picker as an option. I even found someone else to implement this option but his questions went unanswered so he didn't know how to start figuring out how to implement this option: https://github.com/freeplane/freeplane/discussions/1591

I really want this option, even as an add-on or a custom fork (although I really hope to avoid that), but I need some help letting me know how a coder might approach figuring out how to do it. If this option is not in the API, is it even possible to do this with an add-on? Would it require a fork of Freeplane or a big rewrite of major parts of the code? Is it just a matter of knowing the right variable in the code or something? I'm totally ignorant of this sort of thing so I would really appreciate any guidance about how to approach a project to make it possible to do this.

**Comment by macmarrum** [comment]
Freeplane uses Java Swing for its GUI, which provides the component `JFileChooser` → https://docs.oracle.com/javase/8/docs/api/javax/swing/JFileChooser.html

When I copy the example code, from JFileChooser.html, as a script to my mind map (`Tools->Edit script… (Alt+A)`) and I run it, I can see a message naming the file I selected.

```java
import javax.swing.JFileChooser
import javax.swing.filechooser.FileNameExtensionFilter

JFileChooser chooser = new JFileChooser();
FileNameExtensionFilter filter = new FileNameExtensionFilter(
    "JPG & GIF Images", "jpg", "gif");
chooser.setFileFilter(filter);
int returnVal = chooser.showOpenDialog();
if(returnVal == JFileChooser.APPROVE_OPTION) {
   System.out.println("You chose to open this file: " +
        chooser.getSelectedFile().getName());
}
```

![2024-05-09_14,09,00](https://github.com/freeplane/freeplane/assets/7197131/871483da-6ba7-45d1-9d33-94aaedc42a02)

If I were a developer looking to change file choosers in Freeplane, I'd start with 
1. https://duckduckgo.com/?q=Java+Swing+native+file+chooser&ia=web
2. https://github.com/search?q=repo%3Afreeplane%2Ffreeplane%20JFileChooser&type=code

**Reply by quickfold** [reply]
Thank you very much! I'll report back if I am able to make progress.

**Reply by JavaMaxim** [reply]
> Thank you very much! I'll report back if I am able to make progress.

Hey, is this issue fixed, any progress?

**Reply by quickfold** [reply]
I actually spent a lot of time trying to hire someone to make progress on this. I interviewed several people, but all turned the job down for various reasons (e.g. they said the tech stack was too old) or they said it would take at least 20+ hours to do the job, which would cost $400+.

Not sure if you saw the thread where I asked Dimitry's opinion on how to proceed, here: https://github.com/freeplane/freeplane/discussions/2029 . It seems clear that he doesn't support someone else adding this option (@dpolivaev, correct me if I'm wrong), so even if I was willing to spend $400-500 to get this working on Windows, which seems crazy, no one else would be able to benefit from it, because it will never be included in Freeplane official. So it seems like it's not a great use of resources.

If anyone in the community has interest and ability to work with me on this (@macmarrum @EdoFro, @euu2021?) especially if they could do it for a smaller budget, please contact me.

**Reply by dpolivaev** [reply]
@quickfold I thought I answered. Sorry. Anyway: sure: if the option were available, I would happily include it.

**Comment by euu2021** [comment]
Try this script:

```
import java.awt.FileDialog
import java.awt.Frame
import java.io.File

// Create an invisible frame required for the FileDialog
def frame = new Frame()
frame.setVisible(false)

try {
    // Configure the FileDialog for loading a file
    def dialog = new FileDialog(frame, "Select a file to open", FileDialog.LOAD)
    dialog.setVisible(true)

    // Get the selected directory and file
    def directory = dialog.getDirectory()
    def file = dialog.getFile()

    if (file != null) {
        // Construct the full file path
        def filePath = directory + file
        def selectedFile = new File(filePath)

        // Check if the file exists
        if (selectedFile.exists()) {
            // Open the file in Freeplane
            c.mapLoader(selectedFile).withView().getMindMap()
            println "File '${selectedFile.getAbsolutePath()}' opened successfully."
        } else {
            // Display an error message if the file does not exist
            controller.showMessage("Error", "The selected file does not exist:\n${selectedFile.getAbsolutePath()}")
        }
    } else {
        // User canceled the operation
        println "No file was selected."
    }
} catch (Exception e) {
    // Handle any exceptions that occur
    controller.showMessage("Error", "An error occurred while trying to open the file:\n${e.message}")
} finally {
    // Ensure the frame is disposed
    frame.dispose()
}

```

It works for me here, on Windows. It opens the native file chooser.

**Reply by euu2021** [reply]
Wait, nevermind. I just noticed it opens the same default FP File Chooser.

**Reply by quickfold** [reply]
I appreciate the effort!
